File: mycollab-web/src/main/java/com/esofthead/mycollab/module/project/view/ProjectLinkGenerator.java
package com.esofthead.mycollab.module.project.view;

import com.esofthead.mycollab.common.ApplicationProperties;
import com.esofthead.mycollab.common.UrlEncodeDecoder;
import com.esofthead.mycollab.module.project.ProjectContants;

public class ProjectLinkGenerator {

	private static String PARAM_PREFIX = "?url=";

	public static String generateProjectLink(int projectId) {
		return "project/dashboard/" + UrlEncodeDecoder.encode(projectId);
	}

	public static String generateProjectFullLink(Integer projectId) {
		if (projectId == null) {
			return "";
		}
		return ApplicationProperties.getProperty(ApplicationProperties.APP_URL)
				+ PARAM_PREFIX + "project/dashboard/"
				+ UrlEncodeDecoder.encode(projectId);
	}

	public static String generateBugPreviewLink(int projectId, int bugId) {
		return "project/bug/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + bugId);
	}

	public static String generateBugPreviewFullLink(Integer projectId,
			Integer bugId) {
		if (projectId == null || bugId == null) {
			return "";
		}
		return ApplicationProperties.getProperty(ApplicationProperties.APP_URL)
				+ PARAM_PREFIX + generateBugPreviewLink(projectId, bugId);
	}

	public static String generateBugVersionPreviewLink(Integer projectId,
			Integer versionId) {
		return "project/bug/version/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + versionId);
	}

	public static String generateBugComponentPreviewLink(Integer projectId,
			Integer componentId) {
		return "project/bug/component/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + componentId);
	}

	public static String generateMessagePreviewLink(int projectId, int messageId) {
		return "project/message/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + messageId);
	}

	public static String generateMessagePreviewFullLink(Integer projectId,
			Integer messageId) {
		if (projectId == null || messageId == null) {
			return "";
		}
		return ApplicationProperties.getProperty(ApplicationProperties.APP_URL)
				+ PARAM_PREFIX
				+ generateMessagePreviewLink(projectId, messageId);
	}

	public static String generateRiskPreview(Integer projectId, Integer riskId) {
		return "project/risk/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + riskId);
	}

	public static String generateProblemPreviewLink(Integer projectId,
			Integer problemId) {
		return "project/problem/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + problemId);
	}

	public static String generateRiskPreviewFullLink(Integer projectId,
			Integer riskId) {
		if (projectId == null || riskId == null) {
			return "";
		}
		return ApplicationProperties.getProperty(ApplicationProperties.APP_URL)
				+ PARAM_PREFIX + "project/risk/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + riskId);
	}

	public static String generateTaskPreviewFullLink(Integer projectId,
			Integer taskId) {
		if (projectId == null || taskId == null) {
			return "";
		}
		return ApplicationProperties.getProperty(ApplicationProperties.APP_URL)
				+ PARAM_PREFIX + generateTaskPreviewLink(projectId, taskId);
	}

	public static String generateTaskPreviewLink(int projectId, int taskId) {
		return "project/task/task/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + taskId);
	}

	public static String generateTaskGroupPreviewLink(Integer projectId,
			Integer taskgroupId) {
		return "project/task/taskgroup/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + taskgroupId);
	}

	public static String generateTaskGroupPreviewFullLink(Integer projectId,
			Integer taskgroupId) {
		if (projectId == null || taskgroupId == null) {
			return "";
		}
		return ApplicationProperties.getProperty(ApplicationProperties.APP_URL)
				+ PARAM_PREFIX
				+ generateTaskGroupPreviewLink(projectId, taskgroupId);
	}

	public static String generateMilestonePreviewLink(int projectId,
			int milestoneId) {
		return "project/milestone/preview/"
				+ UrlEncodeDecoder.encode(projectId + "/" + milestoneId);
	}

	public static String generateMilestonePreviewFullLink(Integer projectId,
			Integer milestoneId) {
		if (projectId == null || milestoneId == null) {
			return "";
		}
		return ApplicationProperties.getProperty(ApplicationProperties.APP_URL)
				+ PARAM_PREFIX
				+ generateMilestonePreviewLink(projectId, milestoneId);
	}

	public static String generateProjectItemLink(int projectId, String type,
			int typeid) {
		String result = "";

		if (ProjectContants.PROJECT.equals(type)) {
		} else if (ProjectContants.MESSAGE.equals(type)) {
			result = generateMessagePreviewLink(projectId, typeid);
		} else if (ProjectContants.MILESTONE.equals(type)) {
			result = generateMilestonePreviewLink(projectId, typeid);
		} else if (ProjectContants.PROBLEM.equals(type)) {
			result = generateProblemPreviewLink(projectId, typeid);
		} else if (ProjectContants.RISK.equals(type)) {
			result = generateRiskPreview(projectId, typeid);
		} else if (ProjectContants.TASK.equals(type)) {
			result = generateTaskPreviewLink(projectId, typeid);
		} else if (ProjectContants.TASK_LIST.equals(type)) {
			result = generateTaskGroupPreviewLink(projectId, typeid);
		} else if (ProjectContants.BUG.equals(type)) {
			result = generateBugPreviewLink(projectId, typeid);
		} else if (ProjectContants.BUG_COMPONENT.equals(type)) {
			result = generateBugComponentPreviewLink(projectId, typeid);
		} else if (ProjectContants.BUG_VERSION.equals(type)) {
			result = generateBugVersionPreviewLink(projectId, typeid);
		} else if (ProjectContants.STANDUP.equals(type)) {

		}

		return "?url=" + result;
	}
}

File: mycollab-services/src/main/java/com/mycollab/module/project/ProjectLinkGenerator.kt
package com.mycollab.module.project

import com.mycollab.common.GenericLinkUtils
import com.mycollab.common.GenericLinkUtils.URL_PREFIX_PARAM
import com.mycollab.common.UrlEncodeDecoder
import com.mycollab.core.DebugException
import com.mycollab.core.utils.StringUtils

/**
 * @author MyCollab Ltd.
 * @since 1.0
 */
object ProjectLinkGenerator {

    fun generateProjectLink(projectId: Int?): String {
        return "project/dashboard/" + GenericLinkUtils.encodeParam(projectId!!)
    }

    fun generateProjectFullLink(siteUrl: String, projectId: Int?): String {
        return siteUrl + URL_PREFIX_PARAM + generateProjectLink(projectId)
    }

    fun generateTicketDashboardLink(projectId: Int?): String {
        return "project/ticket/dashboard/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateTaskPreviewLink(taskKey: Int?, prjShortName: String): String {
        return String.format("project/task/preview/%s-%d", prjShortName, taskKey)
    }

    fun generateTaskPreviewFullLink(siteUrl: String, taskKey: Int?, prjShortName: String): String {
        return siteUrl + URL_PREFIX_PARAM + generateTaskPreviewLink(taskKey, prjShortName)
    }

    fun generateTaskEditLink(taskKey: Int?, prjShortName: String): String {
        return String.format("project/task/edit/%s-%d", prjShortName, taskKey)
    }

    fun generateMilestonesLink(projectId: Int?): String {
        return "project/milestone/list/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateMilestonePreviewLink(projectId: Int, milestoneId: Int): String {
        return "project/milestone/preview/" + GenericLinkUtils.encodeParam(projectId, milestoneId)
    }

    fun generateMilestonePreviewFullLink(siteUrl: String, projectId: Int, milestoneId: Int): String {
        return siteUrl + URL_PREFIX_PARAM + generateMilestonePreviewLink(projectId, milestoneId)
    }

    fun generateClientPreviewLink(accountId: Int?): String {
        return "project/client/preview/" + UrlEncodeDecoder.encode(accountId)
    }

    fun generateClientPreviewFullLink(siteUrl: String, accountId: Int?): String {
        return siteUrl + URL_PREFIX_PARAM + "project/client/preview/" + UrlEncodeDecoder.encode(accountId)
    }

    fun generatePagesLink(projectId: Int, folderPath: String): String {
        return "project/page/list/" + GenericLinkUtils.encodeParam(projectId, folderPath)
    }

    fun generatePageAdd(projectId: Int, pagePath: String): String {
        return "project/page/add/" + GenericLinkUtils.encodeParam(projectId, pagePath)
    }

    fun generatePageRead(projectId: Int, pagePath: String): String {
        return "project/page/preview/" + GenericLinkUtils.encodeParam(projectId, pagePath)
    }

    fun generatePageEdit(projectId: Int, pagePath: String): String {
        return "project/page/edit/" + GenericLinkUtils.encodeParam(projectId, pagePath)
    }

    fun generateProjectMemberFullLink(siteUrl: String, projectId: Int, memberName: String?): String {
        return if (memberName == null) {
            ""
        } else siteUrl + URL_PREFIX_PARAM + "project/user/preview/" + GenericLinkUtils.encodeParam(projectId, memberName)
    }

    fun generateProjectMemberLink(projectId: Int, memberName: String): String {
        return "project/user/preview/" + GenericLinkUtils.encodeParam(projectId, memberName)
    }

    fun generateRisksLink(projectId: Int?): String {
        return "project/risk/list/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateRiskPreviewLink(projectId: Int, riskId: Int): String {
        return "project/risk/preview/" + GenericLinkUtils.encodeParam(projectId, riskId)
    }

    fun generateRiskPreviewFullLink(siteUrl: String, projectId: Int, riskId: Int): String {
        return siteUrl + URL_PREFIX_PARAM + generateRiskPreviewLink(projectId, riskId)
    }

    fun generateRiskEditLink(projectId: Int?, riskId: Int?): String {
        return "project/risk/edit/" + UrlEncodeDecoder.encode(projectId.toString() + "/" + riskId)
    }

    fun generateRiskAddLink(projectId: Int?): String {
        return "project/risk/add/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateMessageAddLink(projectId: Int?): String {
        return "project/message/add/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateMessagesLink(projectId: Int?): String {
        return "project/message/list/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateMessagePreviewLink(projectId: Int, messageId: Int): String {
        return "project/message/preview/" + GenericLinkUtils.encodeParam(projectId, messageId)
    }

    fun generateMessagePreviewFullLink(siteUrl: String, projectId: Int, messageId: Int): String {
        return siteUrl + URL_PREFIX_PARAM + generateMessagePreviewLink(projectId, messageId)
    }

    fun generateBugComponentPreviewLink(projectId: Int, componentId: Int): String {
        return "project/component/preview/" + GenericLinkUtils.encodeParam(projectId, componentId)
    }

    fun generateBugComponentPreviewFullLink(siteUrl: String, projectId: Int, componentId: Int): String {
        return siteUrl + URL_PREFIX_PARAM + generateBugComponentPreviewLink(projectId, componentId)
    }

    fun generateBugVersionPreviewLink(projectId: Int, versionId: Int): String {
        return "project/version/preview/" + GenericLinkUtils.encodeParam(projectId, versionId)
    }

    fun generateBugVersionPreviewFullLink(siteUrl: String, projectId: Int, versionId: Int): String {
        return siteUrl + URL_PREFIX_PARAM + generateBugVersionPreviewLink(projectId, versionId)
    }

    fun generateBugPreviewLink(bugKey: Int?, prjShortName: String): String {
        return String.format("project/bug/preview/%s-%d", prjShortName, bugKey)
    }

    fun generateBugEditLink(bugKey: Int?, prjShortName: String): String {
        return String.format("project/bug/edit/%s-%d", prjShortName, bugKey)
    }

    fun generateBugsLink(projectId: Int?): String {
        return "project/bug/list/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateBugPreviewFullLink(siteUrl: String, bugKey: Int?, prjShortName: String): String {
        return siteUrl + URL_PREFIX_PARAM + generateBugPreviewLink(bugKey, prjShortName)
    }

    fun generateFileDashboardLink(projectId: Int?): String {
        return "project/file/dashboard/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateTimeReportLink(projectId: Int?): String {
        return "project/time/list/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateInvoiceListLink(projectId: Int?): String {
        return "project/invoice/list/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateRolePreviewLink(projectId: Int, roleId: Int): String {
        return "project/role/preview/" + GenericLinkUtils.encodeParam(projectId, roleId)
    }

    fun generateRolePreviewFullLink(siteUrl: String, projectId: Int, roleId: Int): String {
        return siteUrl + URL_PREFIX_PARAM + generateRolePreviewLink(projectId, roleId)
    }

    fun generateStandupDashboardLink(): String {
        return "project/reports/standup/list/"
    }

    fun generateHoursWeeklyReportLink(): String {
        return "project/reports/weeklytiming/"
    }

    fun generateTimesheetReportLink(): String {
        return "project/reports/timesheet/"
    }

    fun generateUsersWorkloadReportLink(): String {
        return "project/reports/usersworkload/"
    }

    fun generateUsersLink(projectId: Int?): String {
        return "project/user/list/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateProjectSettingLink(projectId: Int?): String {
        return "project/setting/" + UrlEncodeDecoder.encode(projectId)
    }

    fun generateProjectSettingFullLink(siteUrl: String, projectId: Int?): String {
        return siteUrl + URL_PREFIX_PARAM + generateProjectSettingLink(projectId)
    }

    fun generateProjectItemLink(prjShortName: String, projectId: Int, type: String, typeId: String): String {
        var result = ""

        if (typeId == null || StringUtils.isBlank(typeId) || "null" == typeId) {
            return ""
        }

        try {
            if (ProjectTypeConstants.PROJECT == type) {
            } else if (ProjectTypeConstants.MESSAGE == type) {
                result = generateMessagePreviewLink(projectId, Integer.parseInt(typeId))
            } else if (ProjectTypeConstants.MILESTONE == type) {
                result = generateMilestonePreviewLink(projectId, Integer.parseInt(typeId))
            } else if (ProjectTypeConstants.RISK == type) {
                result = generateRiskPreviewLink(projectId, Integer.parseInt(typeId))
            } else if (ProjectTypeConstants.TASK == type) {
                result = generateTaskPreviewLink(Integer.parseInt(typeId), prjShortName)
            } else if (ProjectTypeConstants.BUG == type) {
                result = generateBugPreviewLink(Integer.parseInt(typeId), prjShortName)
            } else if (ProjectTypeConstants.BUG_COMPONENT == type) {
                result = generateBugComponentPreviewLink(projectId, Integer.parseInt(typeId))
            } else if (ProjectTypeConstants.BUG_VERSION == type) {
                result = generateBugVersionPreviewLink(projectId, Integer.parseInt(typeId))
            } else if (ProjectTypeConstants.PAGE == type) {
                result = generatePageRead(projectId, typeId)
            }
        } catch (e: Exception) {
            throw DebugException(String.format("Error generate tooltip%d---%s---%s", projectId, type, typeId), e)
        }

        return "#" + result
    }
}
